src/analysis/normed/ring/seminorm.lean

```diff
@@ -0,0 +1,196 @@
+  p 1 = 1 ↔ p ≠ 0 :=
+begin
+  refine ⟨λ h, ne_zero_iff.mpr ⟨1, by {rw h, exact one_ne_zero}⟩, λ h, _⟩,
+  obtain ⟨x, hx⟩ := ne_zero_iff.mp h,
+  by_cases hp0 : p 1 = 0,
+  { have hx' : p x ≤ 0,
+    { rw ← mul_one x,
+      apply le_trans (p.mul_le x 1) _,
+      rw [hp0, mul_zero], },
+    exact absurd (le_antisymm hx' (p.nonneg x) ) hx },
+  { have h1 : p 1 * 1 ≤ p 1 * p 1,
+    { conv_lhs { rw ← one_mul (1 : R) },
+      convert p.mul_le 1 1,
+      rw mul_one, },
+    rw mul_le_mul_left (lt_of_le_of_ne (p.nonneg _) (ne.symm hp0)) at h1,
+    exact le_antisymm hp h1, }
+end
+
+end ring
+
+end ring_seminorm
+
+/-- The norm of a `non_unital_semi_normed_ring` as a `ring_seminorm`. -/
+def norm_ring_seminorm (R : Type*) [non_unital_semi_normed_ring R] :
+  ring_seminorm R :=
+{ to_fun    := norm,
+  map_zero' := norm_zero,
+  nonneg'   := norm_nonneg,
+  add_le'   := norm_add_le,
+  neg'      := norm_neg,
+  mul_le'   := norm_mul_le }
+
+/-- A function `f : R → ℝ` is a norm on a (nonunital) ring if it is a seminorm and `f x = 0`
+  implies `x = 0`. -/
+structure ring_norm (R : Type*) [non_unital_ring R] extends ring_seminorm R :=
+(ne_zero : ∀ x, x ≠ 0 → 0 < to_fun x)
+
+attribute [nolint doc_blame] ring_norm.to_ring_seminorm
+
+namespace ring_norm
+
+variable [non_unital_ring R]
+
+instance zero_hom_class : zero_hom_class (ring_norm R) R ℝ :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_zero := λ f, f.map_zero' }
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`. -/
+instance : has_coe_to_fun (ring_norm R) (λ _, R → ℝ) := ⟨λ p, p.to_fun⟩
+
+@[simp] lemma to_fun_eq_coe (p : ring_norm R) : p.to_fun = p := rfl
+
+@[ext] lemma ext {p q : ring_norm R} (h : ∀ x, p x = q x) : p = q :=
+fun_like.ext p q h
+
+variable (R)
+
+/-- The trivial norm on a ring `R` is the `ring_norm` taking value `0` at `0` and `1` at every
+  other element. -/
+def trivial_norm [decidable_eq R] : ring_norm R :=
+{ mul_le' := λ x y, begin
+    by_cases h : x * y = 0,
+    { simp only [add_group_seminorm.trivial_norm, if_pos h], apply mul_nonneg;
+      { split_ifs, exacts [le_refl _, zero_le_one] }},
+    { simp only [add_group_seminorm.to_fun_eq_coe, add_group_seminorm.trivial_norm_of_ne_zero h,
+        add_group_seminorm.trivial_norm_of_ne_zero (left_ne_zero_of_mul h),
+        add_group_seminorm.trivial_norm_of_ne_zero (right_ne_zero_of_mul h), mul_one] }
+  end,
+  ne_zero := λ x hx, begin
+    simp only [add_group_seminorm.trivial_norm, if_neg hx],
+    exact zero_lt_one
+  end,
+  .. add_group_seminorm.trivial_norm R }
+
+lemma trivial_norm_of_ne_zero [decidable_eq R] {z : R} (h : z ≠ 0) : trivial_norm R z = 1 :=
+if_neg h
+
+instance [decidable_eq R] : inhabited (ring_norm R) := ⟨trivial_norm R⟩
+
+/-- A nonzero ring seminorm on a field `K` is a ring norm. -/
+def of_ring_seminorm {K : Type*} [field K] (f : ring_seminorm K)
+  (hnt : f ≠ 0) : ring_norm K :=
```

Comment:
Maybe better as `ring_seminorm.to_ring_norm` or `ring_seminorm.ring_norm`, for the sake of dot notation?

Reply:
I've renamed it to `ring_seminorm.to_ring_norm`.